rust/ares/src/codegen.rs

```diff
@@ -0,0 +1,1294 @@
+    };
+    let nock = cg_interpret_inner(context, context_snapshot.virtual_frame, subject, formula);
+    match nock {
+        Ok(noun) => Ok(noun),
+        Err(ae) => Err(exit(context, context_snapshot, ae)),
+    }
+}
+
+/// Fetches or creates codegen code for the subject and formula, then
+/// naively interprets it.
+fn cg_interpret_inner(
+    context: &mut Context,
+    virtual_frame: *const u64,
+    subject: Noun,
+    formula: Noun,
+) -> Result<Noun, ActualError> {
+    // Setup stack for codegen interpretation.
+    // Stack frame layout: [mean trace slow pile dest cont poison registers]
+    // XX update local_noun_pointer() calls with correct constants
+    let mut current_frame: *mut Frame = std::ptr::null_mut();
+    {
+        let pile = cg_pull_pile(context, subject, formula);
+        unsafe { new_frame(context, &mut current_frame, pile, false) };
+    }
+
+    // Load the initial subject to the sire register.
+    {
+        let sire = unsafe { (*(*current_frame).pile.0).sire };
+        register_set(current_frame, sire, subject);
+    }
+
+    // Get the blob, body, and bend nouns from our pile.
+    let mut body: Noun;
+    let mut bend: Noun;
+    {
+        let will = unsafe { (*(*current_frame).pile.0).will };
+        let blob = will
+            .lookup(&mut context.stack, &mut unsafe {
+                (*(*current_frame).pile.0).wish
+            })
+            .unwrap();
+        body = slot(blob, 2).expect("codegen nock error");
+        bend = slot(blob, 3).expect("codegen nock error");
+    }
+
+    loop {
+        if !unsafe { body.raw_equals(D(0)) } {
+            let pole = slot(body, 2).unwrap();
+            body = slot(body, 3).unwrap();
+            match slot(pole, 2).unwrap().as_direct().unwrap().data() {
+                tas!(b"imm") => {
+                    let local = slot(pole, 7).unwrap().as_direct().unwrap().data() as usize;
+                    let value = slot(pole, 6).unwrap();
+                    register_set(current_frame, local, value);
+                }
+                tas!(b"mov") => {
+                    let src = slot(pole, 6).unwrap().as_direct().unwrap().data() as usize;
+                    let dst = slot(pole, 7).unwrap().as_direct().unwrap().data() as usize;
+                    let value = register_get(current_frame, src);
+                    register_set(current_frame, dst, value);
+                }
+                tas!(b"inc") => {
+                    let s = slot(pole, 6).unwrap().as_direct().unwrap().data() as usize;
+                    let d = slot(pole, 7).unwrap().as_direct().unwrap().data() as usize;
+                    let s_value = register_get(current_frame, s);
+                    if let Ok(atom) = s_value.as_atom() {
+                        let value = inc(&mut context.stack, atom);
+                        register_set(current_frame, d, value.as_noun());
+                    } else {
+                        ActualError(Error::Deterministic(D(0)));
```

Comment:
This expression is evaluated *and then thrown away by the `;`*. It will not stop the loop. 
```suggestion
                        break Err(ActualError(Error::Deterministic(D(0))));
```